File: backend_deno_refactor/domain/schedule/event/delete.ts
import type { Result } from "../../lang/result.ts";
import type { RepoError } from "../../repo.ts";
import type { EventNotFound } from "./read.ts";
import type { User } from "../user/model.ts";
import type { EventRepo } from "./repo.ts";
import type { Event } from "./model.ts";
import { ok } from "../../lang/result.ts";
import { eventReadOne } from "./read.ts";

type EventDeleteErrors =
    | RepoError
    | EventNotFound;

export async function eventDelete(
    repo: EventRepo,
    userId: User["id"],
    eventId: Event["id"],
): Promise<Result<void, EventDeleteErrors>> {
    const existingEventResult = await eventReadOne(repo, userId, eventId);
    if (existingEventResult.type === "err") {
        return existingEventResult;
    }
    const deleteResult = await repo.cDelete(eventId);
    if (deleteResult.type === "err") {
        return deleteResult;
    }
    return ok(undefined);
}

File: domain/src/schedule/event/delete.rs
use super::{error::EventErr, model::Event, read::event_read_by_id, repository::EventRepository};

pub async fn event_delete<Repo: EventRepository>(
    repository: &Repo,
    user_id: &str,
    id: &str,
) -> Result<Event, EventErr> {
    let found_event = event_read_by_id(repository, user_id, id).await?;
    repository.delete(&found_event.id).await.map_err(EventErr::DB)?;
    Ok(found_event)
}

#[cfg(test)]
mod tests {
    use crate::{
        database::DBErr,
        schedule::{
            event::{
                error::EventErr, model::stub::event_stub, read::EventIdNotFoundErr,
                repository::stub::EventRepositoryStub,
            },
            user::model::stub::user_stub,
        },
    };

    use super::event_delete;

    #[tokio::test]
    async fn event_delete_ok() {
        assert_eq!(
            event_delete(
                &EventRepositoryStub::of_event(event_stub()),
                &user_stub().id,
                &event_stub().id
            )
            .await,
            Ok(event_stub())
        );
    }

    #[tokio::test]
    async fn event_delete_db_err() {
        assert_eq!(
            event_delete(&EventRepositoryStub::of_db_err(), &user_stub().id, &event_stub().id)
                .await,
            Err(EventErr::DB(DBErr))
        );
    }

    #[tokio::test]
    async fn event_delete_event_id_not_found_err() {
        assert_eq!(
            event_delete(&EventRepositoryStub::of_none(), &user_stub().id, &event_stub().id).await,
            Err(EventErr::EventIdNotFound(EventIdNotFoundErr))
        );
    }
}
